text/0000-contract-metadata.md

```diff
@@ -0,0 +1,117 @@
+Contract metadata allows developers to easily list the methods in a contract and potentially display them to users to
+provide transparency.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there is no direct way for a developer to programmatically list the methods of a contract that is deployed on chain
+because the contract code stored on chain is the compiled wasm code, where the methods name and parameters are already mangled.
+As a result, if developers want to display the methods of a contract to an user of the app to provide transparency,
+especially in the case of financial apps, they have no choice but to hardcode them in the front end, which is suboptimal in many ways.
+Furthermore, if developers want to get the list of methods in some contract for some downstream task like data analysis,
+they have no way of doing so. Contract metadata aims to solve the aforementioned problems by providing a convenient way 
+for developers to list the methods of contracts.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For developers, there will be two main changes:
+- Instead of annotating view and change methods in the `initContract` function in `main.js`,
+they will instead annotate the methods of a contract by decorators.
+More specifically, every method is by default a change method, unless annotated by `@view_method`.
+- Every contract will have a `metadata` method that returns a json that serializes the contract methods in an array.
+For each method, the json serialization is of the form `{"name": <name>, "parameters": [{"name": <param_name1>, "type": <param_type1>, ..}, .. ], "returnType": <return_type>}`.
+
+As an concrete example, suppose we have a contract that maintains a counter on chain:
+
+```typescript
+import { context, storage, near } from "./near";
+
+export function incrementCounterBy(amount: i32 = 1): void {
+  let newCounter = storage.get<i32>("counter") + amount;
+  storage.set<i32>("counter", newCounter)
+  near.log("Counter is now: " + newCounter.toString());
+}
+
+export function decrementCounterBy(amount: i32 = 1): void {
+  let newCounter = storage.get<i32>("counter") - amount;
+  storage.set<i32>("counter", newCounter)
+  near.log("Counter is now: " + newCounter.toString());
+}
+
+@view_method
+export function getCounter(): i32 {
+  return storage.get<i32>("counter");
+}
+```
+
+This contract has two change methods, `incrementCounter` and `decrementCounter`, as well as one view method, `getCounter`.
+In this case, the metadata we want looks like 
+```json
+[
+  {
+    "name": "getCounter",
+    "parameters": [{"name": "amount", "type": "i32"}],
+    "returnType": "i32",
+    "methodType": "view"
+  },
+  {
+    "name": "incrementCounterBy",
+    "parameters": [{"name": "amount", "type": "i32"}],
+    "returnType": "i32",
+    "methodType": "change"
+  },
+  {
+    "name": "decrementCounterBy",
+    "parameters": [{"name": "amount", "type": "i32"}],
+    "returnType": "void",
```

Comment:
I think `null` is more commonly used to denote absence in json, than `void`. I suggest we research the json conventions.

Reply:
Wouldn't this make the type of `returnType` into `string | null` instead of `string` requiring an explicit typecheck (vs a switch case for `"void"`)?

Reply:
Standard JSON serialization/deserialization libraries know how to handle null.

Reply:
I'm talking about the consumers

I can do `switch(returnType)` with cases `"void"`, `"u32"` and `default` whereas with `null` I must do

`if(returnType === null) {}`
`switch(returnType)` with cases `"u32"` and `default`

Reply:
Do we really want to denote absence here? It seems to me that `void` is just a type.

Reply:
We kinda mix type notation from different languages. `void` is JS or C? `f32` is Rust? Let's agree on one thing, I suggest we take the type system from Wasm.

Reply:
It is actually all assemblyscript notation. Wasm does not have a separate type for `void`.

Reply:
Then I suggest we get rid of `void`. The metadata should have a level of generality of Wasm.

Reply:
So what do we do when there is no return type?

Reply:
Also, we can pass non-primitive types as arguments to the contracts, because we can encode then in JSON. What would be a type description for them in metadata?

Reply:
> So what do we do when there is no return type?

Either do not specify "returnType" key in the map, or specify "returnType": null

Reply:
Yes so the plan is to have all the class info listed separately so that binding generation is possible with metadata alone. I'll update the json format to reflect this.

Reply:
I think it's probably better to not specify `returnType` key as it conforms better to wasm and can also save some space.

Reply:
Thanks, serialization of class type is a problem if we want to make it language-agnostic.

Reply:
> I think it's probably better to not specify returnType key as it conforms better to wasm and can also save some space.

Sounds good.